relay/cache/cache.go

```diff
@@ -0,0 +1,30 @@
+package cache
+
+// WeightCalculator is a function that calculates the weight of a key-value pair in a Cache.
+// By default, the weight of a key-value pair is 1. Cache capacity is always specified in terms of
+// the weight of the key-value pairs it can hold, rather than the number of key-value pairs.
+//
+// Unless otherwise noted, Cache implementations are not required to be thread safe.
+type WeightCalculator[K comparable, V any] func(key K, value V) uint64
+
+// Cache is an interface for a generic cache.
+type Cache[K comparable, V any] interface {
+	// Get returns the value associated with the key, and a boolean indicating whether the key was found in the cache.
+	Get(key K) (V, bool)
+
+	// Put adds a key-value pair to the cache. After this operation, values may be dropped if the total weight
+	// exceeds the configured maximum weight. Will ignore the new value if it exceeds the maximum weight
+	// of the cache in and of itself.
+	Put(key K, value V)
+
+	// WithWeightCalculator sets the weight calculator for the cache. May only be called
+	// when the cache is empty. The weight calculator should be an idempotent function that
+	// always returns the same output given the same input.
+	WithWeightCalculator(weightCalculator WeightCalculator[K, V]) error
```

Comment:
It'd be better to drop this method and pass it in from the constructor since it's required to have this calculator

Reply:
done